spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubOperations.java

```diff
@@ -81,12 +81,52 @@
 	 * Send a message to Pub/Sub.
 	 * @param topic the name of an existing topic
 	 * @param payload an object that will be serialized and sent
-	 * @param headers map of String to String headers
 	 * @return the listenable future of the call
 	 */
 	<T> ListenableFuture<String> publish(String topic, T payload,
 			Map<String, String> headers) throws IOException;
 
+	/**
+	 * Send a message to Pub/Sub.
+	 * @param topic the name of an existing topic
+	 * @param payload the message String payload
+	 * @return the listenable future of the call
+	 */
+	ListenableFuture<String> publish(String topic, String payload);
+
+	/**
+	 * Send a message to Pub/Sub.
+	 * @param topic the name of an existing topic
+	 * @param payload the message String payload
+	 * @param charset charset to decode the payload
+	 * @return the listenable future of the call
+	 */
+	ListenableFuture<String> publish(String topic, String payload, Charset charset);
+
+	/**
+	 * Send a message to Pub/Sub.
+	 * @param topic the name of an existing topic
+	 * @param payload the message payload in bytes
+	 * @return the listenable future of the call
+	 */
+	ListenableFuture<String> publish(String topic, byte[] payload);
+
+	/**
+	 * Send a message to Pub/Sub.
+	 * @param topic the name of an existing topic
+	 * @param payload the message payload on the {@link PubsubMessage} payload format
+	 * @return the listenable future of the call
+	 */
+	ListenableFuture<String> publish(String topic, ByteString payload);
+
+	/**
+	 * Send a message to Pub/Sub.
+	 * @param topic the name of an existing topic
+	 * @param payload an object that will be serialized and sent
+	 * @return the listenable future of the call
+	 */
+	<T> ListenableFuture<String> publish(String topic, T payload) throws IOException;
```

Comment:
Why do we need all those overloaded methods, when we have just this one and all the hard conversion logic according the provided payload type can be encapsulated in the implementation details?

Otherwise this is pretty confusing API when I try to use it as auto-completion feature in the IDE.

Reply:
Did that, but kept the publish with string and charset, because there's no elegant way to collapse that one into the new generic one.
Let me know if you disagree.